### PR TITLE
[AERIE-1814] Optimize computations of constraints

### PR DESCRIPTION
**Pull request by adrienmaillard**
* **Tickets addressed:** None
* **Review:** By commit  
* **Merge strategy:** Merge (no squash)  

## Description

@mattdailis and I have noticed that bounds were partially present in constraints but not used. Scheduling often computes resource constraints on very restricted domains but current constraints are computed over the whole simulation results. 

I have added a `bounds` parameter to the `evaluate` method of the `Expression` interface. The previous prototype is still there as a default. This allows for a few optimizations that have been added in the operations of `LinearProfile` and `DiscreteProfile`.

I added a benchmark in which we want to compute a constraint on a very reduced interval (1/1000 of the duration of the profile), the version with time bounding is approx. 100x faster than the one without.  

All of the previously (and applicable) tests were duplicated and modified to test these optimizations.  

## Verification

WIth new tests. 

## Documentation

Look closely to the optimizations in case I missed a corner case. Or a better optimization !

## Future work

None.

### CODE REVIEW

File: constraints/src/main/java/gov/nasa/jpl/aerie/constraints/tree/ViolationsOf.java [hunk 1]
```diff
@@ -18,10 +19,10 @@ public ViolationsOf(Expression<Windows> expression) {
   }
 
   @Override
-  public List<Violation> evaluate(SimulationResults results, Map<String, ActivityInstance> environment) {
-    final var bounds = new Windows(results.bounds);
+  public List<Violation> evaluate(SimulationResults results, final Window bounds, Map<String, ActivityInstance> environment) {
+    final var boundsW = new Windows(bounds);
     final var satisfiedWindows = this.expression.evaluate(results, environment);
```
Comment (mattdailis) [inline constraints/src/main/java/gov/nasa/jpl/aerie/constraints/tree/ViolationsOf.java]: 🔻 Looks like we might have missed one call to `evaluate`.

```suggestion
    final var satisfiedWindows = this.expression.evaluate(results, bounds, environment);
```

File: constraints/src/main/java/gov/nasa/jpl/aerie/constraints/model/LinearProfile.java [hunk 2]
```diff
@@ -58,7 +58,8 @@ public LinearProfile plus(final LinearProfile other) {
               p.valueAt(intersection.start) + o.valueAt(intersection.start),
               p.rate + o.rate
           );
-        }
+        },
+        Window.FOREVER
```
Comment (mattdailis) [inline constraints/src/main/java/gov/nasa/jpl/aerie/constraints/model/LinearProfile.java:62]: 🔻 This change was made in the first commit, but I think it belongs in the second commit.

File: constraints/src/main/java/gov/nasa/jpl/aerie/constraints/tree/Expression.java [hunk 3]
```diff
@@ -2,16 +2,20 @@
...
 import gov.nasa.jpl.aerie.constraints.model.ActivityInstance;
 import gov.nasa.jpl.aerie.constraints.model.SimulationResults;
+import gov.nasa.jpl.aerie.constraints.time.Window;
 
 import java.util.Map;
 import java.util.Set;
 
 public interface Expression<T> {
-  T evaluate(final SimulationResults results, final Map<String, ActivityInstance> environment);
+  T evaluate(final SimulationResults results, final Window bounds, final Map<String, ActivityInstance> environment);
```
Comment (mattdailis) [inline constraints/src/main/java/gov/nasa/jpl/aerie/constraints/tree/Expression.java:11]: 🔶 I do think this is a good approach - one alternative that comes to mind is to have the caller set the `SimulationResults.bounds` field before calling `evaluate`, but I think your way is better.

Tagging @kroffo and @Twisol in case they have any reactions to this.

**Review by mattdailis — CHANGES_REQUESTED** [review]
Looks good to me overall, just one missed call to the wrong `evaluate` method.

> 🔻 Issues to address before merge 
> 🔶 Requests that should not block merge, but should at least be discussed 
> 🔵 Recommendations that can be ignored if desired

File: constraints/src/test/java/gov/nasa/jpl/aerie/constraints/model/DiscreteProfileTimeDomainTest.java [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    ));
+
+    final var other = new DiscreteProfile(List.of(
+        new DiscreteProfilePiece(Window.between(0, Inclusive, 5, Exclusive, SECONDS), SerializedValue.of(true)),
+        new DiscreteProfilePiece(Window.between(5, Inclusive, 10, Exclusive, SECONDS), SerializedValue.of(true)),
+        new DiscreteProfilePiece(Window.between(10, Inclusive, 15, Exclusive, SECONDS), SerializedValue.of(false)),
+        new DiscreteProfilePiece(Window.between(15, Inclusive, 20, Inclusive, SECONDS), SerializedValue.of(false))
+    ));
+
+    final var result = profile.equalTo(other, Window.between(6, 14, SECONDS));
```
Comment (mattdailis) [inline constraints/src/test/java/gov/nasa/jpl/aerie/constraints/model/DiscreteProfileTimeDomainTest.java:31]: I'm a big fan of these test cases :+1:

**Comment by adrienmaillard** [comment]
I have adressed all comments made by @mattdailis in the latest push.

**Review by mattdailis — APPROVED** [review]
Looks good to me

File: constraints/src/test/java/gov/nasa/jpl/aerie/constraints/model/DiscreteProfileTimeDomainTest.java [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+import java.util.List;
+
+import static gov.nasa.jpl.aerie.constraints.Assertions.assertEquivalent;
+import static gov.nasa.jpl.aerie.constraints.time.Window.Inclusivity.Exclusive;
+import static gov.nasa.jpl.aerie.constraints.time.Window.Inclusivity.Inclusive;
+import static gov.nasa.jpl.aerie.merlin.protocol.types.Duration.SECONDS;
+
+public class DiscreteProfileTimeDomainTest {
+
```
Comment (patkenneally) [inline constraints/src/test/java/gov/nasa/jpl/aerie/constraints/model/DiscreteProfileTimeDomainTest.java]: Extra new line

File: constraints/src/test/java/gov/nasa/jpl/aerie/constraints/model/DiscreteProfileTimeDomainTest.java [hunk 6]
```diff
@@ -0,0 +1,103 @@
+package gov.nasa.jpl.aerie.constraints.model;
+
+import gov.nasa.jpl.aerie.constraints.model.DiscreteProfile;
```
Comment (patkenneally) [inline constraints/src/test/java/gov/nasa/jpl/aerie/constraints/model/DiscreteProfileTimeDomainTest.java]: Inellij seems to suggest that these to imports can be removed.

**Review by patkenneally — APPROVED** [review]
